File: Assets/src/model/actors/Body.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public interface IBodyMoveHandler {
  void HandleMove(Vector2Int newPos, Vector2Int oldPos);
}
public interface IBodyTakeAttackDamageHandler {
  void HandleTakeAttackDamage(int damage, int hp, Actor source);
}
public interface ITakeAnyDamageHandler {
  void HandleTakeAnyDamage(int damage);
}
public interface IHealHandler {
  void HandleHeal(int amount);
}
public interface IFloorChangeHandler {
  void HandleFloorChanged(Floor newFloor, Floor oldFloor);
}

[Serializable]
public class Body : Entity {
  private Vector2Int _pos;
  public override Vector2Int pos {
    get => _pos;
    set {
      if (floor == null) {
        _pos = value;
      } else {
        if (floor.tiles[value.x, value.y].CanBeOccupied()) {
          var oldTile = floor.tiles[_pos];
          oldTile.BodyLeft(this);
          _pos = value;
          floor.BodyMoved();
          OnMove(value, oldTile.pos);
          Tile newTile = floor.tiles[_pos];
          newTile.BodyEntered(this);
        } else {
          OnMoveFailed(value);
        }
      }
    }
  }

  /// assumes other is on the same floor
  public void SwapPositions(Actor other) {
    var oldTile = floor.tiles[_pos];
    Tile newTile = floor.tiles[other._pos];
    oldTile.BodyLeft(this);
    newTile.BodyLeft(other);
    _pos = newTile.pos;
    other._pos = oldTile.pos;
    floor.BodyMoved();
    OnMove(_pos, oldTile.pos);
    other.OnMove(oldTile.pos, _pos);
    newTile.BodyEntered(this);
    oldTile.BodyEntered(other);
  }

  public void ChangeFloors(Floor newFloor, Vector2Int newPos) {
    var oldFloor = floor;
    var oldTile = oldFloor?.tiles[_pos];

    oldTile?.BodyLeft(this);
    oldFloor?.Remove(this);

    _pos = newPos;

    // Recompute flag is already set by floor.Remove()
    // floor.BodyMoved();

    // this is NOT a move, so don't trigger OnMove
    // OnMove(newPos, oldTile.pos);

    var newTile = newFloor.tiles[newPos];
    newFloor.Put(this);

    OnFloorChanged(floor, oldFloor);

    newTile.BodyEntered(this);
  }

  public int hp { get; protected set; }
  public int baseMaxHp { get; protected set; }
  public virtual int maxHp => baseMaxHp;
  public override IEnumerable<object> MyModifiers => base.MyModifiers.Append(this.grass?.BodyModifier);
  [field:NonSerialized]
  public event Action OnMaxHPAdded;

  public Body(Vector2Int pos) : base() {
    this.pos = pos;
  }

  protected override void HandleLeaveFloor() {
    tile.BodyLeft(this);
  }

  protected override void HandleEnterFloor() {
    tile.BodyEntered(this);
  }

  protected virtual void OnFloorChanged(Floor newFloor, Floor oldFloor) {
    foreach (var handler in this.Of<IFloorChangeHandler>()) {
      handler.HandleFloorChanged(newFloor, oldFloor);
    }
  }

  /// returns how much it actually healed
  internal int Heal(int amount) {
    if (amount <= 0) {
      Debug.Log("tried healing <= 0");
      return 0;
    }
    amount = Mathf.Clamp(amount, 0, maxHp - hp);
    hp += amount;
    OnHeal(amount);
    return amount;
  }

  public void AddMaxHP(int amount) {
    baseMaxHp += amount;
    OnMaxHPAdded?.Invoke();
  }

  public void Attacked(int damage, Actor source) {
    TakeAttackDamage(damage, source);
  }

  /// Attack damage doesn't always come from an *attack* specifically. For instance,
  /// Snail shells count as attack damage, although they are not an attack action.
  public void TakeAttackDamage(int damage, Actor source) {
    if (IsDead) {
      return;
    }
    damage = Modifiers.Process(this.AttackDamageTakenModifiers(), damage);
    damage = Math.Max(damage, 0);
    source.OnDealAttackDamage(damage, this);
    OnTakeAttackDamage(damage, hp, source);
    TakeDamage(damage, source);
  }

  /// Take damage from any source.
  public void TakeDamage(int damage, Entity source) {
    if (IsDead) {
      return;
    }
    damage = Modifiers.Process(this.AnyDamageTakenModifiers(), damage);
    TakeUnavoidableDamage(damage, source);
  }

  public void TakeUnavoidableDamage(int damage, Entity source) {
    if (IsDead) {
      return;
    }
    damage = Math.Max(damage, 0);
    OnTakeAnyDamage(damage);
    hp -= damage;
    if (hp <= 0) {
      Kill(source);
    }
  }

  public override void Kill(Entity source) {
    /// TODO remove references to this Actor if needed
    if (!IsDead) {
      hp = Math.Max(hp, 0);
      base.Kill(source);
    }
  }

  protected virtual void OnMoveFailed(Vector2Int wantedPos) {}

  private void OnMove(Vector2Int newPos, Vector2Int oldPos) {
    foreach (var handler in this.Of<IBodyMoveHandler>()) {
      handler.HandleMove(newPos, oldPos);
    }
  }

  private void OnTakeAttackDamage(int dmg, int hp, Actor source) {
    foreach (var handler in this.Of<IBodyTakeAttackDamageHandler>()) {
      handler.HandleTakeAttackDamage(dmg, hp, source);
    }
  }

  private void OnTakeAnyDamage(int dmg) {
    foreach (var handler in this.Of<ITakeAnyDamageHandler>()) {
      handler.HandleTakeAnyDamage(dmg);
    }
  }

  private void OnHeal(int amount) {
    foreach (var handler in this.Of<IHealHandler>()) {
      handler.HandleHeal(amount);
    }
  }
}

File: web/src/model/Body.ts
import { Entity } from './Entity';
import { Vector2Int } from '../core/Vector2Int';
import {
  CollisionLayer,
} from '../core/types';
import {
  collectModifiers,
  processModifiers,
  ATTACK_DAMAGE_TAKEN_MOD,
  ANY_DAMAGE_TAKEN_MOD,
  MOVEMENT_LAYER_MOD,
  type IAttackDamageTakenModifier,
  type IAnyDamageTakenModifier,
  type IMovementLayerModifier,
} from '../core/Modifiers';
import { EventEmitter } from '../core/EventEmitter';
import { GameModelRef } from './GameModelRef';

// ─── Handler symbols ───
export const BODY_MOVE_HANDLER = Symbol.for('IBodyMoveHandler');
export const BODY_TAKE_ATTACK_DAMAGE_HANDLER = Symbol.for('IBodyTakeAttackDamageHandler');
export const TAKE_ANY_DAMAGE_HANDLER = Symbol.for('ITakeAnyDamageHandler');
export const HEAL_HANDLER = Symbol.for('IHealHandler');

export interface IBodyMoveHandler {
  handleMove(newPos: Vector2Int, oldPos: Vector2Int): void;
}

export interface IBodyTakeAttackDamageHandler {
  handleTakeAttackDamage(damage: number, hp: number, source: any): void;
}

export interface ITakeAnyDamageHandler {
  handleTakeAnyDamage(damage: number): void;
}

export interface IHealHandler {
  handleHeal(amount: number): void;
}

/**
 * Entity with HP, position movement, and damage pipeline.
 * Port of C# Body.cs.
 */
export class Body extends Entity {
  private _pos: Vector2Int;
  protected _hp: number = 8;
  protected _baseMaxHp: number = 8;

  readonly onMaxHPAdded = new EventEmitter();

  get hp(): number {
    return this._hp;
  }
  protected set hp(value: number) {
    this._hp = value;
  }

  get baseMaxHp(): number {
    return this._baseMaxHp;
  }
  protected set baseMaxHp(value: number) {
    this._baseMaxHp = value;
  }

  get maxHp(): number {
    return this._baseMaxHp;
  }

  /** Override in subclasses for innate movement type (e.g. Flying for Bat). */
  get baseMovementLayer(): CollisionLayer {
    return CollisionLayer.Walking;
  }

  /** Computed from baseMovementLayer + modifiers. */
  get movementLayer(): CollisionLayer {
    const modifiers = collectModifiers<IMovementLayerModifier>(this, MOVEMENT_LAYER_MOD);
    return processModifiers(modifiers, this.baseMovementLayer);
  }

  get pos(): Vector2Int {
    return this._pos;
  }

  set pos(value: Vector2Int) {
    if (!this.floor) {
      this._pos = value;
      return;
    }

    const tile = this.floor.tiles.get(value);
    if (tile && tile.canBeOccupiedBy(this)) {
      const oldTile = this.floor.tiles.get(this._pos)!;
      oldTile.bodyLeft(this);
      const oldPos = this._pos;
      this._pos = value;
      this.floor.bodyMoved();
      GameModelRef.mainOrNull?.emitAnimation({ type: 'move', entityGuid: this.guid, from: oldPos, to: value });
      this.onMove(value, oldPos);
      const newTile = this.floor.tiles.get(this._pos)!;
      newTile.bodyEntered(this);
    } else {
      this.onMoveFailed(value);
    }
  }

  get myModifiers(): Iterable<object | null | undefined> {
    const grassMod = this.grass && 'bodyModifier' in this.grass ? (this.grass as any).bodyModifier : null;
    return [...super.myModifiers, grassMod];
  }

  /** Mark as body for Floor type discrimination */
  readonly _isBody = true;

  constructor(pos: Vector2Int) {
    super();
    this._pos = pos;
  }

  protected handleLeaveFloor(): void {
    this.tile.bodyLeft(this);
  }

  protected handleEnterFloor(): void {
    this.tile.bodyEntered(this);
  }

  heal(amount: number): number {
    if (amount <= 0) return 0;
    amount = Math.min(amount, this.maxHp - this._hp);
    this._hp += amount;
    if (amount > 0) {
      GameModelRef.mainOrNull?.emitAnimation({ type: 'heal', entityGuid: this.guid, to: this._pos, amount });
    }
    this.onHealEvent(amount);
    return amount;
  }

  addMaxHP(amount: number): void {
    this._baseMaxHp += amount;
    this.onMaxHPAdded.emit();
  }

  attacked(damage: number, source: any): void {
    this.takeAttackDamage(damage, source);
  }

  takeAttackDamage(damage: number, source: any): void {
    if (this.isDead) return;
    const mods = collectModifiers<IAttackDamageTakenModifier>(this, ATTACK_DAMAGE_TAKEN_MOD);
    damage = processModifiers(mods, damage);
    damage = Math.max(damage, 0);
    source.onDealAttackDamage(damage, this);
    this.onTakeAttackDamage(damage, this._hp, source);
    this.takeDamage(damage, source);
  }

  takeDamage(damage: number, source: Entity): void {
    if (this.isDead) return;
    const mods = collectModifiers<IAnyDamageTakenModifier>(this, ANY_DAMAGE_TAKEN_MOD);
    damage = processModifiers(mods, damage);
    damage = Math.max(damage, 0);
    this.onTakeAnyDamageEvent(damage);
    if (damage > 0) {
      GameModelRef.mainOrNull?.emitAnimation({ type: 'damage', entityGuid: this.guid, to: this._pos, amount: damage });
    }
    this._hp -= damage;
    if (this._hp <= 0) {
      this.kill(source);
    }
  }

  kill(source: Entity): void {
    if (!this.isDead) {
      this._hp = Math.max(this._hp, 0);
      GameModelRef.mainOrNull?.emitAnimation({ type: 'death', entityGuid: this.guid, to: this._pos });
      super.kill(source);
    }
  }

  setHPDirect(hp: number): void {
    this._hp = hp;
  }

  protected onMoveFailed(_wantedPos: Vector2Int): void {}

  private onMove(newPos: Vector2Int, oldPos: Vector2Int): void {
    for (const handler of collectModifiers<IBodyMoveHandler>(this, BODY_MOVE_HANDLER)) {
      handler.handleMove(newPos, oldPos);
    }
  }

  private onTakeAttackDamage(dmg: number, hp: number, source: any): void {
    for (const handler of collectModifiers<IBodyTakeAttackDamageHandler>(this, BODY_TAKE_ATTACK_DAMAGE_HANDLER)) {
      handler.handleTakeAttackDamage(dmg, hp, source);
    }
  }

  private onTakeAnyDamageEvent(dmg: number): void {
    for (const handler of collectModifiers<ITakeAnyDamageHandler>(this, TAKE_ANY_DAMAGE_HANDLER)) {
      handler.handleTakeAnyDamage(dmg);
    }
  }

  private onHealEvent(amount: number): void {
    for (const handler of collectModifiers<IHealHandler>(this, HEAL_HANDLER)) {
      handler.handleHeal(amount);
    }
  }
}
